File: src/main/java/com/bajdcc/LALR1/semantic/token/TokenBag.java
package com.bajdcc.LALR1.semantic.token;

import com.bajdcc.util.lexer.token.Token;

/**
 * 单词包
 *
 * @author bajdcc
 */
public class TokenBag {

	/**
	 * 保存的单词
	 */
	private Token token = null;

	/**
	 * 保存的对象
	 */
	private Object object = null;

	public TokenBag() {

	}

	public TokenBag(Token token) {
		this.token = token;
	}

	public TokenBag(Object obj) {
		this.object = obj;
	}

	public Token getToken() {
		return token;
	}

	public void setToken(Token token) {
		this.token = token;
	}

	public Object getObject() {
		return object;
	}

	public void setObject(Object object) {
		this.object = object;
	}
}

File: src/main/java/com/bajdcc/LALR1/semantic/token/TokenBag.kt
package com.bajdcc.LALR1.semantic.token

import com.bajdcc.LALR1.grammar.tree.Block
import com.bajdcc.LALR1.grammar.tree.Func
import com.bajdcc.LALR1.grammar.tree.IExp
import com.bajdcc.LALR1.grammar.tree.IStmt
import com.bajdcc.util.lexer.token.Token

/**
 * 单词包
 * [token] 保存的单词
 * [obj] 保存的对象
 * @author bajdcc
 */
data class TokenBag(var token: Token? = null, var obj: Any? = null) {

    fun toExp() = obj!! as IExp
    fun toExps() = (obj!! as List<*>).map { it as IExp }.toMutableList()
    fun toStmt() = obj!! as IStmt
    fun toStmts() = (obj!! as List<*>).map { it as IStmt }.toMutableList()
    fun toTokens() = (obj!! as List<*>).map { it as Token }.toMutableList()
    fun toBlock() = obj!! as Block
    fun toFunc() = obj!! as Func
}